include/zephyr/drivers/gpio.h

```diff
@@ -731,6 +731,9 @@ static inline int z_impl_gpio_pin_configure(const struct device *port,
 	__ASSERT(!((flags & GPIO_INPUT) && !(flags & GPIO_OUTPUT) && (flags & GPIO_SINGLE_ENDED)),
 		 "Input cannot be enabled for 'Open Drain', 'Open Source' modes without Output");
 
+	__ASSERT((flags & (GPIO_INPUT | GPIO_OUTPUT)) != (GPIO_INPUT | GPIO_OUTPUT),
+		 "Input and Output cannot be set simultaneously");
```

Comment:
Having a GPIO configured as both input and output is generally valid. It is up to the GPIO driver to reject such a configuration if not supported by the given GPIO controller.